### ISSUE TITLE
What happens when a parametrised test fails?

### ISSUE THREAD

**Issue by hadley** [google/patrick#20]
i.e. what does the error + traceback look like, and how do you quickly get to a minimal reprex that you can re-run to debug the problem?

**Comment by michaelquinn32** [comment]
Patrick dynamically forms test cases and reports failures as if you wrote the test cases.

So for a test that modifies this: https://github.com/google/patrick/blob/b6b4430c97a940e9a4ed4867f5155735875aab67/tests/testthat/test-with_parameters.R#L73

```r
with_parameters_test_that(
  "Data frames can be passed to cases:",
  {
    result <- rlang::as_function(FUN)(input)
    testthat::expect_equal(result, out)
  },
  .cases = tibble::tribble(
    ~.test_name, ~FUN, ~input, ~out,
    "times", ~ .x * 2, 2, 5,
    "plus", ~ .x + 3, 3, 6
  )
)
```

We see a failure like this:

```
[ FAIL 1 | WARN 1 | SKIP 0 | PASS 17 ]

══ Failed tests ════════════════════════════════════════════════════════════════
── Failure (test-with_parameters.R:77:5): Data frames can be passed to cases: times ──
`result` (`actual`) not equal to `out` (`expected`).

  `actual`: 4
`expected`: 5
Backtrace:
    ▆
 1. ├─rlang::eval_tidy(code, args)
 2. └─testthat::expect_equal(result, out) at test-with_parameters.R:77:5

[ FAIL 1 | WARN 1 | SKIP 0 | PASS 17 ]
 ```

What could we do to make that output clearer? Is this an issue of docs or of something in the tests?

**Comment by michaelquinn32** [comment]
We don't have a great story around reprexes, since the code is generated by rlang::eval_tidy() right now in a way that isn't accessible to the test author.
https://github.com/google/patrick/blob/b6b4430c97a940e9a4ed4867f5155735875aab67/R/with_parameters.R#L163

**Comment by hadley** [comment]
I almost wonder if a code generation approach would be better — i.e. do the interpolation and then render to disk before running the tests. Then you can easily reproduce failures.

However that would require some thinking about what exactly you're capturing in cases: is it the values or the expressions used to create those values?

**Comment by michaelquinn32** [comment]
I'm trying to apply patterns from the python parameterized library.
https://pypi.org/project/parameterized/

And cases are mostly about variable values. 
https://github.com/google/googletest/blob/main/docs/advanced.md#value-parameterized-tests

We could do more to report the failing values with the error message. All of this could be injected into the test case description.
https://github.com/google/patrick/blob/b6b4430c97a940e9a4ed4867f5155735875aab67/R/with_parameters.R#L159

**Comment by hadley** [comment]
I'm thinking more like the example on the home page:

```R
numeric_data <- 1:10

with_parameters_test_that("Data is successfully converted:", {
    input <- convert(test_data)
    expect_type(input, type)
  },
  test_data = list(numeric_data, character_data),
  type = c("double", "character"),
  .test_name = type
)
```

If you were to generate the failing test code, would you want it to look like this:

```R
test_that("Data is successfully converted:", {
  input <- convert(numeric_data)
  expect_type(input, "double")
})
```

or like

```R
test_that("Data is successfully converted:", {
  input <- convert(1:10)
  expect_type(input, "double")
})
```

(with the caveat that it would actually have the 1:10 vector inlined into the AST).

It'll get more challenging with complex data types which don't yield such nice code when deparsing the AST.

**Comment by hadley** [comment]
(To give you a bit more context about my feedback, patrick looks super interesting, and I think it'd be worth considering including directly in testthat (assuming you were interested). But the big challenge is handling failing test cases. I've been experiencing the struggle lately with [DBItest](https://dbitest.r-dbi.org). It's clearly super useful to be able to package up tests for cross-package interfaces, but it's currently very painful to debug when a test fails. To me, the biggest possible value add for a system like patrick over for-loops or functional programming, is really thinking through the affordances when a test fails.)

**Comment by michaelquinn32** [comment]
Thanks!

I am a little reluctant to write intermediate outputs to disk. We rely on a CI environment where writing local temp files isn't possible.

Could we handle this with a ReprexReporter, descending from this:
https://testthat.r-lib.org/reference/Reporter.html

It seems that a debug reporter already exists and might be similar:
https://github.com/r-lib/testthat/blob/main/R/reporter-debug.R

The reprex reporter would:

*  Print out local variables in the environment (since that where cases are set within patrick)
*  Then print out the body of the test

It might be something like this:

````
══ Failed tests ════════════════════════════════════════════════════════════════
── Failure (test-with_parameters.R:77:5): Data frames can be passed to cases: times ──
`result` (`actual`) not equal to `out` (`expected`).

  `actual`: 4
`expected`: 5
Backtrace:
    ▆
 1. ├─rlang::eval_tidy(code, args)
 2. └─testthat::expect_equal(result, out) at test-with_parameters.R:77:5

Reprex:
```
# Local variables
test_data <- 1:10
type <- "double"

# Test body
input <- convert(test_data)
expect_type(input, type)
```
````

**Comment by hadley** [comment]
The problem with a reporter is that you'd have to opt-in to it somehow.

I think it might work for `with_parameters_test_that()` to catch expectation failures/errors then rethrow with the reprex added in the text of the condition.

**Comment by michaelquinn32** [comment]
Another option for handling parameters / iteration from the Python unittest library is subtests.
https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests

This might work better in testthat, if there is a way to isolate behavior within the subtests.

**Comment by hadley** [comment]
@michaelquinn32 ooh that's super interesting, I made [a note in testthat](https://github.com/r-lib/testthat/issues/2063) so I don't forget about it.

**Comment by michaelquinn32** [comment]
These are mostly my own notes, but I want to get them written down.

It is possible to capture the fully substituted test case code. You need to follow the suggestion here and use `do.call("substitute", ...)`.
https://stackoverflow.com/questions/26638746/how-can-i-do-partial-substitution-in-r

It's not that helpful though, because we still wrap test_that.

*  Passing fully-substituted code is a little weird because test_that calls substitute() itself
*  We could try to use tryCatch() but test_that handles errors itself

What we really want is to better report what's happening in the test case on failure. For that, we can use some existing patterns for labeling test cases and extend the descriptions. This can be gnarly if there are a lot of parameters, so for now I'm mostly documenting this behavior in the README.

*  https://github.com/google/patrick/pull/28
* And https://github.com/google/patrick/blob/b356a297bb162481d15d713199c222b8f6bc8a54/R/with_parameters.R#L143

**Comment by MichaelChirico** [comment]
@hadley it might be a good time to pick this back up with the recent flurry of {testthat} activity -- is this something to consider for the next release? I see https://github.com/r-lib/testthat/issues/2063 was filled.

**Comment by hadley** [comment]
@MichaelChirico yes, I think so. The primary change is that `describe()`, `it()`, and `test_that()` now all work the same way, by explicitly managing a stack of test names that are pasted together to generate a failure message. 

This makes it easier to use loops and functions:

```R
power_function <- function(x, n) {
  if (n < 0) stop("Negative exponents not supported")
  if (x == 0 && n == 0) stop("0^0 is undefined")
  x^n
}

test_that("power_function works for multiple bases and exponents", {
  test_cases <- data.frame(
    x = c(2, 3, 4, 5),
    n = c(2, 2, 2, 2),
    expected = c(4, 9, 16, 25)
  )
  
  for (i in seq_len(nrow(test_cases))) {
    test_that(paste("x =", test_cases$x[i], "n =", test_cases$n[i]), {
      expect_equal(
        power_function(test_cases$x[i], test_cases$n[i]),
        test_cases$expected[i]
      )
    })
  }
})

# Helper function for testing
test_power <- function(x, n, expected) {
  test_that(paste0("power_function(", x, ", ", n, ") equals ", expected), {
    expect_equal(power_function(x, n), expected)
  })
}

# Test many combinations
test_that("power combinations work", {
  test_power(2, 3, 8)
  test_power(5, 2, 25)
  test_power(10, 0, 1)
  test_power(-3, 2, 9)
})
```

I think this circles back to the original idea here, i.e. the test name gives you the details of the failure message.
